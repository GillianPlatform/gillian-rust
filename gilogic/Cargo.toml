[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }