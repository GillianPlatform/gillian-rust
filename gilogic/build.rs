fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
