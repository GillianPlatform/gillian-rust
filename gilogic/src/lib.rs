#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
