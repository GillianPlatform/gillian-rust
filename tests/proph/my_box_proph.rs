//@check-pass
extern crate gilogic;

struct MyBox<T> {
    ptr: *mut T,
}

use gilogic::{
    macros::{assertion, predicate, specification},
    prophecies::Ownable,
};

impl<T: Ownable> Ownable for MyBox<T> {
    type RepresentationTy = T::RepresentationTy;
    #[predicate]
    fn own(self, model: Self::RepresentationTy) {
        assertion!(|ptr, v| (self == MyBox { ptr }) * (ptr -> v) * v.own(model))
    }
}

impl<T: Ownable> MyBox<T> {

    #[specification(
        forall sm, vm .
            requires { self.own(sm) * v.own(vm) }
        exists rm, foo: T .
            ensures {   ret.own(rm) * $rm == vm$ }
    )]
    // #[requires(|sm: T::RepresentationTy, vm: T::RepresentationTy| self.own(sm) * v.own(vm) )]
    // #[ensures(|rm: T::RepresentationTy, foo: T|  ret.own(rm) * $rm == vm$ )]
    fn write(self, v: T) -> Self {
        unsafe { *self.ptr = v };
        self
    }

    // #[requires(|sm: (T::RepresentationTy, T::RepresentationTy), vm: T::RepresentationTy| self.own(sm) * v.own(vm))]
    // #[ensures(|rm: T::RepresentationTy| $sm.1 == vm$ * $sm.0 == vm$)]
    // fn update(&mut self, v: T) {
    //     unsafe { *self.ptr = v };
    // }
}
