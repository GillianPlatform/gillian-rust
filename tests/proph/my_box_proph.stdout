#begin_init_data
{"MyBox":["Struct","ReprRust",[["ptr",["Ptr",{"mut":true,"ty":["Param",0]}]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_pcy.gil", "i__lang_pcy.gil";

pred "<MyBox<T> as gilogic::prophecies::Ownable>::own"(+pty_T0, +self, model):
  (#self == {{ #ptr }}) * <value>(l-nth(#ptr, 0i), l-nth(#ptr, 1i), pty_T0; #v)
  * "<T as gilogic::prophecies::Ownable>::own"(pty_T0, #v, #model)
  * (model == #model)
  * (self == #self);


abstract pred "<T as gilogic::prophecies::Ownable>::own"(+pty_T0, +self, model);


spec "MyBox::<T>::write"(pty_T0, self, v)
[[ "<MyBox<T> as gilogic::prophecies::Ownable>::own"(#pty_T0, #self, #sm)
  * "<T as gilogic::prophecies::Ownable>::own"(#pty_T0, #v, #vm)
  * (v == #v)
  * (self == #self)
  * (pty_T0 == #pty_T0) ]]
[[ "<MyBox<T> as gilogic::prophecies::Ownable>::own"(#pty_T0, ret, #rm)
  * <observation>(((#rm) = (#vm)); ) ]]
normal

proc "MyBox::<T>::write"(pty_T0, self, v) {
        m_1self := self;
        g_0 := m_1self;
        m_1self := [alloc]({{ "adt", "MyBox", {{ pty_T0 }} }});
        u := [store_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "adt", "MyBox", {{ pty_T0 }} }}, g_0);
        m_2v := v;
  bb0:  g_1 := m_2v;
        m_4 := g_1;
        goto bb1;
  bb1:  g_2 := m_4;
        g_4 := [load_value](l-nth(m_1self, 0i), l+(l-nth(m_1self, 1i), {{ {{ "f", 0i, {{ "adt", "MyBox", {{ pty_T0 }} }} }} }}), {{ "ptr", true, pty_T0 }}, true);
        g_3 := l-nth(g_4, 0i);
        u := [store_value](l-nth(g_3, 0i), l-nth(g_3, 1i), pty_T0, g_2);
        m_3 := {{}};
        goto bb3;
  bb3:  g_6 := [load_value](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "adt", "MyBox", {{ pty_T0 }} }}, false);
        g_5 := l-nth(g_6, 0i);
        m_0 := g_5;
        goto bb4;
  bb4:  g_7 := m_0;
        ret := g_7;
        goto rlab;
  rlab: u := [free](l-nth(m_1self, 0i), l-nth(m_1self, 1i), {{ "adt", "MyBox", {{ pty_T0 }} }});
        return
};

